src/Telemetry/Microsoft.DotNet.AzureDevOpsTimelineTests/MockHttpMessageHandler.cs

```diff
@@ -0,0 +1,61 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.AzureDevOpsTimeline.Tests
+{
+    public class MockHttpMessageHandler : HttpMessageHandler
+    {
+        private readonly Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> _sendAsync;
+
+        public int RequestCancelledCount { get; private set; }
+
+        public MockHttpMessageHandler(Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> sendAsync)
+        {
+            _sendAsync = sendAsync;
+        }
+
+        public static MockHttpMessageHandler Create(string message)
+        {
+            return new MockHttpMessageHandler(async (request, cancellationToken) =>
+            {
+                var responseMessage = new HttpResponseMessage(HttpStatusCode.OK)
+                {
+                    Content = new StringContent(message)
+                };
+
+                return await Task.FromResult(responseMessage);
+            });
+        }
+
+        public static MockHttpMessageHandler CreateWithCustomHttpStatusCode(HttpStatusCode httpStatusCode)
+        {
+            return new MockHttpMessageHandler(async (request, cancellationToken) =>
+            {
+                var responseMessage = new HttpResponseMessage(httpStatusCode);
+
+                return await Task.FromResult(responseMessage);
+            });
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            try
+            {
+                await Task.Delay(1000, cancellationToken);
```

Comment:
If it's not necessary, I would omit real time waits in tests. It is very unpredictable. What do you need this for?

Can you maybe have here some synchronization task source completion instead to block on this in the test?

Reply:
We are trying to simulate a slow internet connection here, checking if the cancellation is handled properly

Reply:
Yeah, let's not do that. Unit tests get executed in parallel and this would bring a lot of flakiness into the equation. Also the test will then take a long time.

You can do the trick I showed in the Tuesday talk with task completions to create a barrier where the code will stop hanging and you have full control.

Reply:
You can also use a "TaskCompletionSource" into the mock.  Assert that before you've called it, it doesn't trigger, and then after completing it, it does.  That should be able to run without depending on any delays and not have any flakiness, if this scenario is something you really need to test.